### ISSUE TITLE
Feature request: create/scan TOTP codes

### ISSUE THREAD

**Issue by keepassium** [keepassium/KeePassium#24]
It should be possible to add TOTP codes to entries (manually or by scanning a QR code).

(Related to #23)

**Comment by igorkulman** [comment]
No ETA for this feature?

**Comment by keepassium** [comment]
Not in the next few months, unfortunately...

**Comment by goetzc** [comment]
@keepassium Have you though on having something like https://issuehunt.io to get bounties for features requests?

**Comment by keepassium** [comment]
@goetzc , thanks for the suggestion. After checking their website just now, I either miss their point or it does not fit.

As far as I understand, their idea is to collect donations to fund the development of features. (Similarly to BountySource.) If so, this would not help. The issue is not the money, but me being the bottleneck.

- To keep control over project's licensing, I need to maintain full copyright on it. That is, any contributor would have to either transfer their copyright to me or publish their contribution under a permissive (MIT/BSD/CC0) license. 
  - Copyright transfer forms are the antithesis of fun, so I don't expect anyone to contribute this way.
  - Managing a mix of licenses sounds like a pain. (This class is GPL licensed, but this method in it is MIT licensed, but this line in it was modified by another author...)
- External contributions entail a certain management overhead. What if contribution quality is low? Do I ask to improve it? Do I reject it and alienate the person? Do something else? Whatever is the decision, I'd waste hours contemplating it instead of working on the project :)
- Last but not least. Free contributions imply an ethical problem. KeePassium has a commercial side and monetizing someone's good will just does not feel right. (A $10 bounty is pretty much a "free contribution", too.)

Bottom line: if you know a good Swift developer available for freelancing, please drop me a line :)

**Comment by igorkulman** [comment]
> 
> Bottom line: if you know a good Swift developer available for freelancing, please drop me a line :)

So you are basically saying if I create this feature, because I would really like to have it, you cannot accept a PR for free, but you would buy it from me? 😀

**Comment by keepassium** [comment]
> So you are basically saying if I create this feature, because I would really like to have it, you cannot accept a PR for free, but you would buy it from me? 😀

Haha, sounds counterintuitive, doesn't it? :) I was tempted to say "yes", but there are caveats.

Unlike contributed code, "commissioned" code clarifies many formalities. It's like those [1-Euro houses](https://edition.cnn.com/travel/article/italy-one-euro-home-websites/index.html) sold in Italy. They could simply give away the buildings to anyone who asks. But then it would be unclear who owns the building, by what right, who has to pay taxes on it, etc. Buying the house for the symbolic €1 makes it a standard deal. 

Same here: if the code is "commissioned" rather than "contributed":

- It becomes "work for hire", so copyright belongs to the client
- The client can define the requirements (code style, UI layout, etc)
- Completion is measured against the requirements
- If requirements change, the client pays for ongoing work
- Clearer accounting for both parties

This is different from "buying a PR" — at least in the requirements part — but yeah, I would pay you to implement this :)

**Comment by goetzc** [comment]
@keepassium thanks for your explanation, I complete understand your point. 

I got this idea from a similar, but desktop project I use, [KeePassXC](https://github.com/keepassxreboot/keepassxc). They also have bounties for devs to fix/implement some issues. I think the bounty is given when project owners consider the contribution good enough to get merged.

Maybe you can use some ideas from how [OniVim](https://github.com/onivim/oni2#license) managed they license, on how to manage open source contributions on a mixed (license or copyright-wise) project like this. Not sure if that helps :)
 
BTW, I think BountySource is abandoned.

Anyway, I love this app, thank you!

**Comment by keepassium** [comment]
Thanks, @goetzc. 

Bounties make sense for KeePassXC, as they are a bona fide FOSS project: open to contributions and funded by donations. My focus is on building a commercial project, where open source is required for transparency. For a "commercial project" ambition, buying PRs seems more appropriate than symbolic bounties.

> Maybe you can use some ideas from how OniVim managed they license, on how to manage open source contributions on a mixed (license or copyright-wise) project like this. Not sure if that helps :)

Thanks for the link! It seems they require contributors both [sign a CLA](https://onivim.github.io/docs/for-developers/contributing#submitting-a-pull-request) and make contributions available under the MIT license. That's a lot of legalese hoops to jump through... Hopefully it works for them :)
